### ISSUE TITLE
Use with S3 self hosted version

### ISSUE THREAD

**Issue by DrewRay** [codedellemc/rexray#849]
# Summary
How do we use the rexray plugin with a custom URL for the S3 provider.  in the s3fs FUSE layer we can do this with the [URL parameter](https://github.com/s3fs-fuse/s3fs-fuse/wiki/Fuse-Over-Amazon#url-defaulthttpss3amazonawscom)  but this appears to be missing from the way rexray allows you to specify, it forces me to use "region" which appends to the URL.  

i have tried --s3fsOptions url=https://ooss.example.com in the command line and the below in by config.yml but neither seems to work.

s3fs:
  cmd: s3fs
  options:
  - url=https://ooss.example.com
  - use_path_request_style

## Version

```shell
# rexray version
REX-Ray
-------
Binary: /usr/bin/rexray
Flavor: client+agent+controller
SemVer: 0.9.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: 2a7458dd90a79c673463e14094377baf9fc8695e
Formed: Wed, 03 May 2017 16:38:11 CDT

libStorage
----------
SemVer: 0.6.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: fa055d6da595602715bdfd5541b4aa6d4dcbcbd9
Formed: Wed, 03 May 2017 16:36:11 CDT
```

## Expected Behavior
rexray volume create test2 --size=1 -l debug -v -fsType s3fs --s3fsDisablePathStyle --s3fsOptions url=https://ooss.example.com  --s3fsAccessKey MASK --s3fsSecretKey MASK 

should replace the host with an appropriate URL path.

## Actual Behavior
not working  URL rremains amazonAWS

## Steps To Reproduce
try to use a self-hosted URL with S3fs

you will see the below in the log output

```
---[ REQUEST POST-SIGN ]-----------------------------
GET / HTTP/1.1
Host: s3.amazonaws.com
User-Agent: aws-sdk-go/1.6.18 (go1.7.5; linux; amd64)
Authorization: AWS4-HMAC-SHA256 Credential=1Z5rngkyPndRjBWra5XC/20170512/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=9eb81f913124f3cd5bb804c8ec6a6f807b281a32cedd0e960851fa5fa5d009cb
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170512T204934Z
Accept-Encoding: gzip
```

**Comment by akutz** [comment]
Hi @DrewRay,

Have you attempted using a config file at `/etc/rexray/config.yml`?

```yaml
s3fs:
  options:
  -url=https://ooss.example.com/
  -accessKey=mask
  -secretKey=mask
```

**Comment by akutz** [comment]
FWIW, I think a bug was introduced in 0.9 that is causing this to fail via the CLI. I am going to file this issue as a bug. I believe it's because the S3FS driver isn't registering its configuration with the registry in a callback as it now needs to do.

**Comment by akutz** [comment]
Thank you for finding this out!

**Comment by DrewRay** [comment]
Yes I had the issue using the config.yml as well.  Sorry it got lost in the original post.  I will send my config when I get back to my computer later.

**Comment by sebglon** [comment]
I'm interesting with this and How can i pass this url on docker plugin?

**Comment by PhilippeChepy** [comment]
Any news about fixing this issue ?

**Comment by fredix** [comment]
Hi, I need to use minio instead of amazon S3. What about url option ?

**Comment by codenrhoden** [comment]
The `0.9.2` version of this plugin went into Docker Hub about 3 weeks ago, which contained the fix for passing `S3FS_OPTIONS`. That should allow a `url` option to be passed. The README on the hub doesn't show the new options, but they are listed [here](http://rexray.readthedocs.io/en/stable/user-guide/docker-plugins/#configuration_3).

Has anyone given that a try?

**Comment by srhb** [comment]
@codenrhoden I'm not sure how to even pass the options in. Neither of these prevent the plugin from attempting and failing to talk with Amazon instead of my local server:

    docker install rexray/s3fs:0.9.2 S3FS_OPTIONS="-o url=http://myserver:8080" (...keys...)
    docker install rexray/s3fs:0.9.2 S3FS_OPTIONS="url=http://myserver:8080" (...keys...)

**Comment by dahendel** [comment]
We have submitted the following pull request which allows you to specify a custom s3 endpoint in your config.yml [https://github.com/codedellemc/rexray/pull/986](url). You do still need to add the -url=customEndpoint in the options key so that when s3fs command is executed it is passed properly.

Example Config:

```
libstorage:
  service: s3fs
  integration:
    volume:
      operations:
        create:
          default:
            size: 100
        mount:
          rootPath: /rexray 
s3fs:
  accessKey: AccessKey
  secretKey: SecretKey
  endpoint: CustomS3Url
  options:
  - url=CustomS3Url
  - use_path_request_style
  - allow_other
  - nonempty
```

**Comment by srhb** [comment]
Linking up with the PR so this issue gets updated :-)

#986

**Comment by clintkitson** [comment]
@dahendel Is your example right where `s3fs.endpoint` is specified while the `s3fs.options` `url` is also specified?

**Comment by dahendel** [comment]
@clintkitson yes. The reason for both is that thewhen the s3 storage driver executes fuse-s3fs it passes the options in to mount the bucket. If you do not pass the url option it throws an error because s3fs still tries to call amazon s3.

We created the endpoint key because when the storage driver creates the s3 connection using the aws.Config struct there is an option for a custom endpoint that was not originally passed.

```

**Comment by clintkitson** [comment]
That makes sense to me. I am just wondering if there is any reason the URL might be different.

**Comment by dahendel** [comment]
I can't think of a really good reason it would be different, they should always be the same. We could set the url= option if the endpoint key is set the same way we do it in the driver config. Then, we wouldn't have to set the url twice.

**Comment by dahendel** [comment]
I also tested adding the below code in the s3fs_executer.go init to add the -url opt to d.opts if endpoint is set. This works, but if -url is set in the options key it still works but is then passed twice to the s3fs command.

```
if ep := d.config.GetString(s3fs.ConfigS3FSEndpoint); ep != "" {
		d.opts = append(d.opts, fmt.Sprintf("url=%s", ep))
	}
```
I have not pushed that change in the PR. We would obviously want to sanitize the options to strip out the additional url option.

**Comment by gcavalcante8808** [comment]
Hello Folks, any news on this? We want to use DigitalOceanSpaces a storage for some simple config files instead of configMaps or docker swarm configs.

**Comment by clintkitson** [comment]
Hello @gcavalcante8808,

I believe this is possible today. See the configurable options https://github.com/thecodeteam/rexray/blob/master/.docker/plugins/s3fs/config.json#L55-L109

`docker install rexray/s3fs S3FS_OPTIONS="-o url=http://myserver:8080" S3FS_ENDPOINT="" S3FS_ACCESSKEY="" S3FS_SECRETKEY=""`

**Comment by doncicuto** [comment]
Hi @gcavalcante8808  and @clintkitson,
if you want to install the plugin using a 3rd party like minio or wasabi, you should use S3FS_OPTIONS and specidy the URL that you use in the S3FS_ENDPOINT as @clintkitson mentioned, but there's a typo it should be S3FS_OPTIONS="url=http://..." removing the -o. For example: 

`docker plugin install rexray/s3fs S3FS_OPTIONS="allow_other,use_path_request_style,nonempty,url=http://127.0.0.1:9000" S3FS_ENDPOINT="http://127.0.0.1:9000" S3FS_ACCESSKEY="xxxxxxxxxxxxxxxx" S3FS_SECRETKEY="xxxxxxxxxxxxxxxxxx" --alias s3fs --grant-all-permissions`

Hope that helps!

**Comment by piscis** [comment]
@doncicuto THANK's that seams to work and created the bucket's. I'm going to test in depth now!

**Comment by clintkitson** [comment]
Thanks @doncicuto

**Comment by bweston92** [comment]
@doncicuto any idea how to get it working with Docker Compose too? Thanks.

**Comment by doncicuto** [comment]
@bweston92 I haven't tried I'll let you know if I have some time to try it.

**Comment by gcavalcante8808** [comment]
@doncicuto I confirm that using the specified options, it works flawlessly with Minio. Thank you!

**Comment by bweston92** [comment]
Gabriel you ok to submit a PR with details to do this to the minio/minfs
repo?

**Comment by llitfkitfk** [comment]
@doncicuto 
digital ocean space tested and success!

BTW: where can I find all options for S3FS_OPTIONS?

**Comment by doncicuto** [comment]
@llitfkitfk Not sure if this answer will help you, but for S3FS_OPTIONS I think you may have two sources: 

- [The FUSE README](https://github.com/fuse4x/fuse) where you have options like nonempty and allow_other explained.
- S3FS Fuse man page from [GitHub](https://github.com/s3fs-fuse/s3fs-fuse/blob/35261e6dba7a3071b382d15b7bc31a2fb2b799f2/doc/man/s3fs.1) (raw) or a [compiled online version.](http://manpages.ubuntu.com/manpages/xenial/man1/s3fs.1.html)

**Comment by xkrt** [comment]
Looks like I forced to specify endpoint and keys on `docker plugin install rexray/s3fs` stage, not on `docker volume create -d rexray/s3fs` stage. Why is this? So I can't use more than one S3 service on single docker machine (for ex. AWS S3 and Minio)?

Look to docker documentation [example](https://docs.docker.com/storage/volumes/#initial-set-up), there we install sshfs plugin single time without any configuration and can use it to create multiple volumes with multiple target machines.

**Comment by clintkitson** [comment]
Unfortunately this is the docker model re plugins. A plug-in can manually
be created and named differently to get by this however.

**Comment by xkrt** [comment]
@clintkitson I believe docker plugin model is completely ok. I should install s3fs plugin single time and use it to create multiple volumes that map to multiple s3 endpoints/buckets. This is strange idea for me "install plugin multiple times with different aliases per endpoint" and not in the spirit of docker plugin-model. IMHO.

**Comment by clintkitson** [comment]
Yes but the question is whether docket enables a single plugin to address
multiple endpoints. Based on only accepting a volume name, I don’t think so.

**Comment by xkrt** [comment]
@clintkitson please, look at docker docs [example](https://docs.docker.com/storage/volumes/#initial-set-up):
```
$ docker plugin install --grant-all-permissions vieux/sshfs
$ docker volume create --driver vieux/sshfs \
  -o sshcmd=test@node2:/home/test \
  -o password=testpassword \
  sshvolume
```
I can pass opts on `docker volume create` stage, where I can point this volume on specific node and path in case of sshfs.

So, for s3fs I believe it can be something like:
```
$ docker plugin install --grant-all-permissions rexray/s3fs
$ docker volume create --driver rexray/s3fs \
  -o S3FS_OPTIONS="url=http://my-minio:9000,use_path_request_style" \
  -o S3FS_ENDPOINT="http://my-minio:9000" \
  -o S3FS_BUCKET=bucket1 \
  -o S3FS_ACCESSKEY="MY_MINIO_ACCESS_KEY" \
  -o S3FS_SECRETKEY="MY_MINI_SECRET_KEY" \
  minio_bucket1
```

**Comment by bigred94** [comment]
I'm trying @doncicuto solution with minio and seems to be close, I can create a volume and it creates a bucket. If I attach that volume to a container and create a file in the volume of the container a data/ directory is created in the bucket that then has the file in it. The issues is when I'm connecting to an existing bucket it does not show its contents. It's like it's looking for a data directory but none of my existing buckets are structured like that. How do I get the volume to be the base of the bucket?


Edit: Figured it out. 
Had to specify root path: 
`docker plugin set <plugin> LIBSTORAGE_INTEGRATION_VOLUME_OPERATIONS_MOUNT_ROOTPATH="/"`
